katsdpcontroller/product_config.py

```diff
@@ -31,63 +82,1196 @@ class _Sensor(ABC):
+        output_channels = config.get('output_channels')
+        if output_channels is not None:
+            output_channels = tuple(output_channels)
+        return cls(
+            name, src_streams,
+            output_channels=output_channels,
+            parameters=config.get('parameters', {}),
+            min_time=config.get('min_time', DEFAULT_SPECTRAL_MIN_TIME)
+        )
+
+
+STREAM_CLASSES: Mapping[str, Type[Stream]] = {
+    'cbf.antenna_channelised_voltage': AntennaChannelisedVoltageStream,
+    'cbf.tied_array_channelised_voltage': TiedArrayChannelisedVoltageStream,
+    'cbf.baseline_correlation_products': BaselineCorrelationProductsStream,
+    'sim.cbf.antenna_channelised_voltage': SimAntennaChannelisedVoltageStream,
+    'sim.cbf.tied_array_channelised_voltage': SimTiedArrayChannelisedVoltageStream,
+    'sim.cbf.baseline_correlation_products': SimBaselineCorrelationProductsStream,
+    'cam.http': CamHttpStream,
+    'sdp.vis': VisStream,
+    'sdp.beamformer': BeamformerStream,
+    'sdp.beamformer_engineering': BeamformerEngineeringStream,
+    'sdp.cal': CalStream,
+    'sdp.flags': FlagsStream,
+    'sdp.continuum_image': ContinuumImageStream,
+    'sdp.spectral_image': SpectralImageStream
 }
 
 
+class Configuration:
+    def __init__(self, *,
+                 options: Options,
+                 simulation: Simulation,
+                 streams: Iterable[Stream]) -> None:
+        self.options = options
+        self.simulation = simulation
+        self.streams = list(streams)
+        self._by_class: Dict[Type[Stream], List[Stream]] = {}
+        for stream in streams:
+            self._by_class.setdefault(type(stream), []).append(stream)
+
+    def by_class(self, stream_cls: Type[_S]) -> Sequence[_S]:
+        return self._by_class.get(stream_cls, [])    # type: ignore
+
+    @classmethod
+    async def from_config(cls, config: Mapping[str, Any]) -> 'Configuration':
+        _validate(config)
+        config = _upgrade(config)
+        options = Options.from_config(config.get('config', {}))
+        simulation = Simulation.from_config(config.get('simulation', {}))
+        # First get the cam.http stream, so that sensors can be extracted
+        cam_http: Optional[CamHttpStream] = None
+        stream_configs = {**config.get('inputs', {}), **config.get('outputs', {})}
+        for name, stream_config in stream_configs.items():
+            if stream_config['type'] == 'cam.http':
+                cam_http = CamHttpStream.from_config(options, name, stream_config, [], {})
+                break
+
+        # Extract sensor values. This is pulled into a separate block rather
+        # than done as we construct each stream so that one day it can be
+        # optimised to make a single sensor_values call for all sensors.
+        sensors: Dict[str, Dict[str, Any]] = {name: {} for name in stream_configs}
+        if cam_http:
+            client = katportalclient.KATPortalClient(str(cam_http.url), None)
```

Comment:
Does the client not need any disposal once we are done with it ?

Reply:
I looked into that. There is a `disconnect` method, but that's only meaningful if it's established a websocket connection for streaming updates. For HTTP requests it uses a tornado AsyncHTTPClient, which has a singleton per IOLoop, so doesn't need cleanup.